source/mir/random/ndvariable.d

```diff
@@ -0,0 +1,293 @@
+    if (isFloatingPoint!T)
+{
+    ///
+    enum isRandomVariable = true;
+
+    private size_t dim;
+    private AlphaParams alpha;
+
+    /++
+    Params:
+        alpha = (array of) concentration parameters
+    Constraints: `alpha[i] > 0`
+    +/
+    this(AlphaParams alpha)
+    {
+        assert(alpha.length >= 1);
+        for(size_t i = 0; i < alpha.length; ++i)
+            assert(alpha[i] > T(0));
+
+        this.dim = alpha.length - 1;
+        this.alpha = alpha;
+    }
+
+    ///
+    void opCall(G)(ref G gen, T[] result)
+    {
+        assert(result.length == dim + 1);
+        opCall(gen, result.sliced);
+    }
+
+    ///
+    void opCall(G, SliceKind kind, Iterator)(ref G gen, Slice!(kind, [1], Iterator) result)
+        if (isSaturatedRandomEngine!G)
+    {
+        assert(result.length == dim + 1);
+        for(size_t i = 0; i < result.length; ++i)
+            result[i] = GammaVariable!T(alpha[i])(gen);
+        result[] /= result.sum!"kbn";
+    }
+}
+
+///
+unittest
+{
+    auto gen = Random(unpredictableSeed);
+    auto rv = DirichletVariable!double([1.0, 5.7, 0.3]);
+    double[3] x;
+    rv(gen, x);
+    assert(x[0] >= 0 && x[1] >= 0 && x[2] >= 0);
+    assert(fabs(x[0] + x[1] + x[2] - 1) < 1e-10);
+}
+
+/++
+ Compute Cholesky decomposition in place. Only accesses lower/left half of
+ the matrix. Returns false if the matrix is not positive definite.
+ +/
+private bool cholesky(SliceKind kind, Iterator)(Slice!(kind, [2], Iterator) m)
+    if(isFloatingPoint!(DeepElementType!(typeof(m))))
+{
+    alias dotm = reduce!"a - b * c"; // note the `-`
+    assert(m.length!0 == m.length!1);
+
+    /* this is a straight-forward implementation of the Cholesky-Crout algorithm
+    from https://en.wikipedia.org/wiki/Cholesky_decomposition#Computation */
+    for(size_t i = 0; i < m.length!0; ++i)
+    {
+        for(size_t j = 0; j < i; ++j)
+            m[i,j] = dotm(m[i,j], m[i,0..j], m[j,0..j]) / m[j,j];
+        m[i,i] = dotm(m[i,i], m[i,0..i], m[i,0..i]);
+        if(!(m[i,i] > 0)) // this catches nan's as well
+            return false;
+        m[i,i] = sqrt(m[i,i]);
+    }
+    return true;
+}
+
+/++
+$(Multivariate normal distribution).
++/
+struct MultivariateNormalVariable(T, MuParams = T[], SigmaParams = ContiguousMatrix!T)
+    if(isFloatingPoint!T)
+{
+    ///
+    enum isRandomVariable = true;
+
+    private size_t dim;
+    private MuParams mu; // mean vector
+    private SigmaParams sigma; // cholesky decomposition of covariance matrix
+    private NormalVariable!T norm;
+
+    /++
+    Constructor computes the Cholesky decomposition of `sigma` in place without
+    memory allocation. Furthermore it is assumed to be a symmetric matrix, but
+    only the lower/left half is actually accessed.
+
+    Params:
+        mu = mean vector
+        sigma = covariance matrix
+
+    Constraints: sigma has to be positive-definite
+    +/
+    this(MuParams mu, SigmaParams sigma)
```

Comment:
`this(MuParams mu, SigmaParams sigma, bool computeCholesky = true)`

Reply:
I like this suggestion as sometimes one passes the Cholesky. It might make sense to re-name `private SigmaParams sigma;` to `private SigmaParams cholSigma;`

I would also recommend adding the following constructor
`this(SigmaParams sigma, bool computeCholesky = true)`
that assumes that mu is a zero vector.